### PR TITLE
mysql_user: priv parameter can also be a dictionary

### LINKED ISSUE #57533: List support for mysql_user privs
##### SUMMARY

Allow lists or dictionaries for privilege specifications, for enhanced readability.

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME

- mysql_user module

##### ADDITIONAL INFORMATION

Currently, mysql_user privileges are specified using a single string and a forward slash as separator between different sets of privileges.

However, when a single user requires multiple privilege sets, this quickly becomes unreadable and hard to maintain:
```yaml
- mysql_user:
    name: user1
    privs: "db1.*:ALL/db2.*:SELECT,INSERT/db3.table1,table2:SELECT/db3.table3:SELECT,DELETE"
```

I propose list support for increased legibility:
```yaml
- mysql_user:
    name: user1
    privs:
      - "db1.*:ALL"
      - "db2.*:SELECT,INSERT"
      - "db3.table1,table2:SELECT"
      - "db3.table3:SELECT,DELETE"
```

Alternatively, a dictionary would further increase clarity and maintainability by separating objects from privileges:
```yaml
- mysql_user:
    name: user1
    privs:
      "db1.*": "ALL"
      "db2.*": "SELECT,INSERT"
      "db3.table1,table2": "SELECT"
      "db3.table3": "SELECT,DELETE"
```

### PR DESCRIPTION
**Pull request by Andersson007**
##### SUMMARY
Fixes: https://github.com/ansible/ansible/issues/57533

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
```mysql_user```

### CODE REVIEW

**Comment by Andersson007** [comment]
@meersjo @bmalynovytch is this implementation ok ?

**Comment by Andersson007** [comment]
ready_for_review

**Comment by Andersson007** [comment]
@bmalynovytch could you please take a look

**Comment by bmalynovytch** [comment]
@Andersson007 hadn't enough time to give it a look.
I looked at it quickly, sound great.
One question though: why not using only one single parameter and allow it to be from different types ?
If it is a dict => use your new code
If it is a string => use as is

**Comment by Andersson007** [comment]
@bmalynovytch firstly it crossed my find but it won't pass sanity checks after and i wouldn't like to add anything to ```ignore.txt```. I suffered enough trying to clean it out:)

**Comment by Andersson007** [comment]
@gundalow what do you think is better - the new option (i add by this PR) or the current one which will check a type of an argument + we will add new rows to ignore.txt?
If it's allowable to use ignore.txt in this case, i can change the PR like @bmalynovytch suggests.

**Comment by Andersson007** [comment]
@felixfontein suggested using ```raw``` type
@bmalynovytch i'll change the PR, thanks!

**Comment by Andersson007** [comment]
@felixfontein i recalled one thing. The value can counaint commas, won't it be interpreted as a list then?
But we need to have only string or dict there.

**Comment by ffdybuster** [comment]
@Andersson007 the list splitting should only be done if `type=list`; for `type=raw` you simply get the value as it is from the YAML (after conversion to JSON and reading it back, that is). You have to make sure that other types than string and `dict` are rejected by the module.

**Comment by ffdybuster** [comment]
(And make sure that values which have been correctly converted to string before are still treated as strings, like ints or floats. No idea if that's true for this option though...)

**Comment by Andersson007** [comment]
@bmalynovytch , changed
ready_for_review

**Comment by meersjo** [comment]
Looks like exactly what I wanted, thank you :-)

I like that it's now the same parameter; I think that makes it a bit less confusing to use. Will be happy to convert my playbooks once this is in production!

**Review by bmalynovytch — APPROVED** [review]
LGTM 👍 
shipit

**Comment by Andersson007** [comment]
@meersjo @bmalynovytch @felixfontein thanks for the idea and suggestion about ```row```

File: lib/ansible/modules/database/mysql/mysql_user.py [hunk 1]
```diff
@@ -649,6 +659,20 @@ def privileges_grant(cursor, user, host, db_table, priv):
...
+
+    Args:
+        priv (dict): Dict of privileges that needs to be converted to string.
+
+    Returns:
+        priv (str): String representation of input argument.
+    """
+    priv_list = ['%s:%s' % (key, val) for key, val in iteritems(priv)]
+
+    return '/'.join(priv_list)
```
Comment (felixfontein) [inline lib/ansible/modules/database/mysql/mysql_user.py:674]: Does the order matter? If yes, it's probably better to use
```suggestion
    return '/'.join(sorted(priv_list))
```

Comment (Andersson007) [reply lib/ansible/modules/database/mysql/mysql_user.py:674]: I think the order doesn’t matter

**Pull request merged**